File: backend/src/modules/pdt/pdt.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like } from 'typeorm';
import { Pdt } from './entities/pdt.entity';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';
import { KamailioRpcService } from '../../common/services/kamailio-rpc.service';

@Injectable()
export class PdtService {
  constructor(
    @InjectRepository(Pdt)
    private pdtRepository: Repository<Pdt>,
    private kamailioRpcService: KamailioRpcService,
  ) {}

  async findAll(dto: PaginationDto): Promise<PaginatedResult<Pdt>> {
    const { page, limit, keyword } = dto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) where.prefix = Like(`%${keyword}%`);
    const [items, total] = await this.pdtRepository.findAndCount({
      where, skip, take: limit, order: { sdomain: 'ASC', prefix: 'ASC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async create(data: Partial<Pdt>): Promise<Pdt> {
    return this.pdtRepository.save(this.pdtRepository.create(data));
  }

  async update(id: number, data: Partial<Pdt>): Promise<Pdt> {
    const item = await this.pdtRepository.findOne({ where: { id } });
    if (!item) throw new NotFoundException('记录不存在');
    Object.assign(item, data);
    return this.pdtRepository.save(item);
  }

  async remove(id: number): Promise<void> {
    await this.pdtRepository.delete(id);
  }

  async reload(): Promise<void> {
    await this.kamailioRpcService.call('pdt.reload');
  }

  async list(): Promise<any> {
    return this.kamailioRpcService.call('pdt.list');
  }
}

File: kam-admin-module-routing/src/main/java/io/kamailio/admin/modules/pdt/service/PdtService.java
package io.kamailio.admin.modules.pdt.service;

import io.kamailio.admin.common.RpcTimeouts;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.common.service.KamailioRpcService;
import io.kamailio.admin.modules.pdt.entity.Pdt;
import io.kamailio.admin.modules.pdt.repository.PdtRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class PdtService {
    private final PdtRepository repository;
    private final KamailioRpcService kamailioRpc;

    public PdtService(PdtRepository repository, KamailioRpcService kamailioRpc) {
        this.repository = repository;
        this.kamailioRpc = kamailioRpc;
    }

    public PaginatedResult<Pdt> findAll(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? repository.findByPrefixContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("sdomain", "prefix")))
                : repository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("sdomain", "prefix")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    @Transactional
    public Pdt create(Pdt data) { return repository.save(data); }

    @Transactional
    public Pdt update(Integer id, Pdt data) {
        var item = repository.findById(id).orElseThrow(() -> new RuntimeException("记录不存在"));
        if (data.getSdomain() != null) item.setSdomain(data.getSdomain());
        if (data.getPrefix() != null) item.setPrefix(data.getPrefix());
        if (data.getDomain() != null) item.setDomain(data.getDomain());
        return repository.save(item);
    }

    @Transactional
    public void remove(Integer id) { repository.deleteById(id); }

    public void reload() { kamailioRpc.pdtReload().block(RpcTimeouts.DEFAULT_BLOCK); }
    public Object list() { return kamailioRpc.pdtList().block(RpcTimeouts.DEFAULT_BLOCK); }
}
